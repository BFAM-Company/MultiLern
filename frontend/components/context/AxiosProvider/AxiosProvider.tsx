import React, { ReactNode, createContext, useContext } from "react";
import axios, { AxiosInstance } from "axios";
import { AuthContext } from "../AuthContext/AuthContext";
import createAuthRefreshInterceptor from "axios-auth-refresh";
import AsyncStorage from "@react-native-async-storage/async-storage";
// import * as Keychain from "react-native-keychain";

interface AuthState {
  accessToken: string | null;
  refreshToken: string | null;
}

interface AxiosContextProps {
  authAxios: AxiosInstance;
  publicAxios: AxiosInstance
}

const AxiosContext = createContext<AxiosContextProps>({
  authAxios: axios.create({
    baseURL: "http://localhost:3001/auth",
  }),
  publicAxios: axios.create({
    baseURL: "http://localhost:3001/",
  })
});

const { Provider } = AxiosContext;

const AxiosProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const authContext = useContext(AuthContext);

  const authAxios = axios.create({
    baseURL: "http://localhost:3001/auth",
  });

  const publicAxios = axios.create({
    baseURL: "http://localhost:3001/",
  });

  authAxios.interceptors.request.use(
    (config) => {
      if (!config.headers.Authorization) {
        config.headers.Authorization = `Bearer ${authContext?.authState.accessToken}`;
      }

      return config;
    },
    (error) => {
      return Promise.reject(error);
    }
  );

  const refreshAuthLogic = (failedRequest: any) => {
    const data = {
      refreshToken: authContext?.authState.refreshToken,
    };

    const options = {
      method: "POST",
      data,
      url: "http://localhost:3001/auth/refresh",
    };

    return axios(options)
      .then(async (tokenRefreshResponse: any) => {
        failedRequest.response.config.headers.Authorization =
          "Bearer " + tokenRefreshResponse.data;

        await AsyncStorage.setItem('auth',JSON.stringify({
            accessToken: tokenRefreshResponse.data,
            refreshToken: authContext?.authState.refreshToken,
        }))

        authContext?.setAuthState({
          ...authContext.authState,
          accessToken: tokenRefreshResponse.data,
        });
        
        return Promise.resolve();
      })
      .catch(async (e: any) => {
        authContext?.setAuthState({
          accessToken: null,
          refreshToken: null,
          authenticated: false
        });
        await AsyncStorage.clear()
      });
  };

  createAuthRefreshInterceptor(authAxios, refreshAuthLogic, {});

  return (
    <Provider
      value={{
        authAxios,
        publicAxios,
      }}
    >
      {children}
    </Provider>
  );
};

export { AxiosContext, AxiosProvider };
