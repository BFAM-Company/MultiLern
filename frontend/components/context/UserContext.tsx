import React, { createContext, useState, ReactNode } from "react";

export interface UserDataState {
  nickname: string;
  email: string;
  avatar: any;
  isLogged: boolean
}

export interface UserDataContextType {
  userData: UserDataState | undefined;
  setUserData: (userData: UserDataState) => void;
}


const UserDataContext = createContext<UserDataContextType | null>(null);
const { Provider } = UserDataContext;

const UserDataProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [userData, setUserData] = useState<UserDataState>();

  return (
    <Provider
      value={{
        userData,
        setUserData
      }}
    >
      {children}
    </Provider>
  );
};

export { UserDataContext, UserDataProvider };
