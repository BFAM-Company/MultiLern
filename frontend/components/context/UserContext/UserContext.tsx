import React, { createContext, useState, ReactNode } from "react";

export interface UserDataState {
  id: number,
  nickname: string;
  email: string;
  avatar: any;
}

export interface UserDataContextType {
  userData: UserDataState | undefined;
  setUserData: (userData: UserDataState) => void;
}


const UserDataContext = createContext<UserDataContextType | null>(null);
const { Provider } = UserDataContext;

const UserDataProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [userData, setUserData] = useState<UserDataState>();

  return (
    <Provider
      value={{
        userData,
        setUserData
      }}
    >
      {children}
    </Provider>
  );
};

export { UserDataContext, UserDataProvider };
