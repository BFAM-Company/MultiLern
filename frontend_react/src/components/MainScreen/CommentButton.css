.modal {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  display: flex;
  justify-content: center;
  align-items: flex-start;
  padding-top: 20px;
}

.modal-container {
  width: 80%;
  background-color: white;
  border-radius: 20px;
  padding: 20px;
  overflow-y: auto;
}

.close-button-container {
  display: flex;
  justify-content: flex-end;
}

.close-button {
  background: transparent;
  border: none;
}

.hint-text {
  margin-top: 20px;
  font-size: 20px;
  font-weight: 900;
  color: rgb(45, 45, 55);
}

.small-hint-text {
  font-size: 16px;
  color: gray;
}

.input,
textarea {
  width: 90%;
  padding: 10px;
  border-radius: 10px;
  background-color: #fff;
  border: 1px solid #ccc;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
}

.image-gallery {
  display: flex;
  flex-wrap: wrap;
  gap: 10px;
}

.upload-button,
.image-button {
  width: 70px;
  height: 70px;
  border: 1px solid black;
  border-radius: 5px;
  padding: 10px;
  margin-top: 10px;
}

.submit-button {
  width: 300px;
  height: 55px;
  border-radius: 20px;
  margin: 20px;
  background: linear-gradient(45deg, rgb(45, 45, 55), rgb(100, 100, 110));
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
}

.submit-button:hover {
  background: rgb(100, 100, 110);
}
