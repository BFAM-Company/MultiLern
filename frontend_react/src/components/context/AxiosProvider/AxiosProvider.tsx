import React, { ReactNode, createContext, useContext } from "react";
import axios, { AxiosInstance } from "axios";
import { AuthContext } from "../AuthContext/AuthContext";
import createAuthRefreshInterceptor from "axios-auth-refresh";

interface AuthState {
  accessToken: string | null;
  refreshToken: string | null;
}

interface AxiosContextProps {
  authAxios: AxiosInstance;
  publicAxios: AxiosInstance;
}

// Kontekst Axios
const AxiosContext = createContext<AxiosContextProps>({
  authAxios: axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}/auth`,
    baseURL: `${import.meta.env.VITE_API_URL}/auth`,
  }),
  publicAxios: axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}`,
    baseURL: `${import.meta.env.VITE_API_URL}`,
  }),
});

const { Provider } = AxiosContext;

const AxiosProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const authContext = useContext(AuthContext);

  const authAxios = axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}/auth`,
    baseURL: `${import.meta.env.VITE_API_URL}/auth`,
  });

  const publicAxios = axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}`,
    baseURL: `${import.meta.env.VITE_API_URL}`,
  });

  // Interceptor dodający nagłówki autoryzacji
  authAxios.interceptors.request.use(
    (config) => {
      if (authContext?.authState.accessToken) {
        config.headers.Authorization = `Bearer ${authContext.authState.accessToken}`;
      }
      return config;
    },
    (error) => Promise.reject(error)
  );

  // Funkcja obsługująca odświeżanie tokenu
  const refreshAuthLogic = (failedRequest: any) => {
    const data = {
      refreshToken: authContext?.authState.refreshToken,
    };

    return axios
      // .post(`${process.env.REACT_APP_API_URL}/auth/refresh`, data)
      .post(`${import.meta.env.VITE_API_URL}/auth/refresh`, data)
      .then((tokenRefreshResponse) => {
        const newAccessToken = tokenRefreshResponse.data;

        // Ustawienie nowego tokenu w localStorage
        localStorage.setItem(
          "auth",
          JSON.stringify({
            accessToken: newAccessToken,
            refreshToken: authContext?.authState.refreshToken,
          })
        );

        // Aktualizacja stanu kontekstu
        authContext?.setAuthState((prevState) => ({
          ...prevState,
          accessToken: newAccessToken,
        }));

        // Ponowienie oryginalnego żądania
        failedRequest.response.config.headers.Authorization = `Bearer ${newAccessToken}`;
        return Promise.resolve();
      })
      .catch(() => {
        // Wylogowanie w przypadku niepowodzenia
        authContext?.setAuthState({
          accessToken: null,
          refreshToken: null,
          authenticated: false,
        });
        localStorage.clear();
      });
  };

  // Dodanie interceptora obsługującego odświeżanie tokenu
  createAuthRefreshInterceptor(authAxios, refreshAuthLogic);

  return (
    <Provider value={{ authAxios, publicAxios }}>
      {children}
    </Provider>
  );
};

export { AxiosContext, AxiosProvider };
