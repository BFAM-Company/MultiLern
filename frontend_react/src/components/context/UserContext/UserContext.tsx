import React, { createContext, useState, ReactNode } from "react";

// Typy danych uÅ¼ytkownika
export interface UserDataState {
  id: number;
  nickname: string;
  email: string;
  avatar: string | null;
}

export interface UserDataContextType {
  userData: UserDataState | undefined;
  setUserData: (userData: UserDataState) => void;
}

// Tworzenie kontekstu
const UserDataContext = createContext<UserDataContextType | null>(null);

const UserDataProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [userData, setUserData] = useState<UserDataState | undefined>(undefined);

  return (
    <UserDataContext.Provider
      value={{
        userData,
        setUserData,
      }}
    >
      {children}
    </UserDataContext.Provider>
  );
};

export { UserDataContext, UserDataProvider };
