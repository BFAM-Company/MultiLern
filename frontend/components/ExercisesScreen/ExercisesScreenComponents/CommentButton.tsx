import { LinearGradient } from 'expo-linear-gradient';
import React, { useContext, useState } from 'react'
import {StyleSheet, TouchableOpacity, Text, Modal, Dimensions, View} from 'react-native'
import { AxiosContext } from '../../context/AxiosProvider/AxiosProvider';
import { UserDataContext } from '../../context/UserContext/UserContext';





interface PostContentProps{
    id:number,
    handleRefresh: any,
}

const windowWidth = Dimensions.get('window').width;
  const windowHeight = Dimensions.get('window').height;
function CommentButton({id, handleRefresh}: PostContentProps) {
  const userContext = useContext(UserDataContext)
  const {publicAxios, authAxios} = useContext(AxiosContext);
  const [modalVisible, setModalVisibile] = useState<boolean>(false)

     
  const handleClick = () =>{
    setModalVisibile(true)
  }

  const handleSubmit = async() =>{
        const event = new Date();

        await publicAxios.post(`/posts/comment`, {
            title: 'Moja odpowiedz',
            content: 'masz rozwiÄ…zanie',
            date: event.toISOString(),
            parentPostId: id,
            userId: userContext?.userData?.id,
            }, {
              headers: {
                'Content-Type': 'application/json'
              }
          })

        await handleRefresh()
      }


  return (
    <>
        <Modal
            visible={modalVisible}
            style={{width:windowWidth, height:windowHeight}}
        >
            <View
                style={styles.modalContainer}
            >

            </View> 
        </Modal>
        <TouchableOpacity
            onPress={handleClick}
            style={styles.button}
        >
            <LinearGradient
                colors={['rgb(45,45,55)', 'rgb(45,45,55)', 'rgb(100,100,110)']}
                style={styles.linearGradient}
            >
                <Text style={{color:'white', fontWeight:'900', fontSize:22}}>Odpowiedz</Text>
            </LinearGradient>
        </TouchableOpacity>
    </>
    
  );
}




const styles = StyleSheet.create({
    button:{
        width:300,
        height:55,
        borderRadius:20,
        margin:20,
    },
    linearGradient:{
        width:'100%',
        height:'100%',
        display:'flex',
        justifyContent:'center',
        borderRadius:20,
        alignItems:'center',
    },
    modalContainer:{
        width:'100%',
        height:'100%',
        display:'flex',
        alignItems:'center',
        justifyContent:'space-between',
    },
})

export default CommentButton;