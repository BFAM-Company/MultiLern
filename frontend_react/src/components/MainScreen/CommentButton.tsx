import React, { useContext, useState } from 'react';
import './CommentButton.css';
import * as ImagePicker from 'react-image-picker-editor'; // Use a suitable library for image picking
import { UserDataContext } from '../context/UserContext/UserContext';
import { AxiosContext } from '../context/AxiosProvider/AxiosProvider';

interface PostContentProps {
  id: number;
  handleRefresh: any;
}

interface IImages {
  images: {
    create: {
      img: string;
    };
  };
}[];

function CommentButton({ id, handleRefresh }: PostContentProps) {
  const userContext = useContext(UserDataContext);
  const { publicAxios, authAxios } = useContext(AxiosContext);
  const [modalVisible, setModalVisible] = useState<boolean>(false);

  const [title, onChangeTitle] = useState<string>('');
  const [content, onChangeContent] = useState<string>('');
  const [images, setImages] = useState<IImages[]>([]);

  const pickImage = async () => {
    // Handle image picking for the web (use a file input or an image picker library)
    const result = await ImagePicker.pickImage({ quality: 0.1 });

    if (result) {
      const image: IImages = {
        images: {
          create: {
            img: result.base64,
          },
        },
      };
      setImages((prevState) => [...prevState, image]);
    }
  };

  const handleClick = () => {
    setModalVisible(true);
  };

  const removeImage = (imageId: number) => {
    const updatedImages = images.filter((_, index) => index !== imageId);
    setImages(updatedImages);
  };

  const handleCommentSubmit = async () => {
    const event = new Date();

    await publicAxios.post(
      `/posts/comment`,
      {
        title: title,
        content: content,
        date: event.toISOString(),
        parentPostId: id,
        images: images,
        userId: userContext?.userData?.id,
      },
      {
        headers: {
          'Content-Type': 'application/json',
        },
      }
    );
    setModalVisible(false);
    await handleRefresh();
  };

  return (
    <>
      {modalVisible && (
        <div className="modal" style={{ width: '100%', height: '100%' }}>
          <div className="modal-container">
            <div className="close-button-container">
              <button onClick={() => setModalVisible(false)} className="close-button">
                <img src="/assets/close-icon-bold.png" alt="Close" style={{ width: 30, height: 30 }} />
              </button>
            </div>
            <h2 className="hint-text">Dodaj tytuł do swojej odpowiedzi!</h2>
            <p className="small-hint-text">
              Może ułatwić to późniejsze znalezienie twojej odpowiedzi innym użytkownikom
            </p>
            <input
              style={{ width: '90%' }}
              onChange={() => onChangeTitle}
              value={title}
              placeholder="tytuł(opcjonalny)..."
            />
            <h2 className="hint-text">Dodaj treść swojej odpowiedzi</h2>
            <p className="small-hint-text">Pomóż innym. Staraj się wyjaśnić rozwiązanie i swój tok myślenia!</p>
            <textarea
              style={{ width: '90%', minHeight: 400 }}
              onChange={(e) => onChangeContent(e.target.value)}
              value={content}
              placeholder="Rozwiązanie..."
            />
            <h2 className="hint-text">Dodaj obrazy do swojej odpowiedzi</h2>
            <div className="image-gallery">
              {images.map((image: any, index) => (
                <button key={index} onClick={() => removeImage(index)} className="image-button">
                  <img style={{ width: 100, height: 100, opacity: 0.7, borderRadius: 5 }} src={image.images.create.img} />
                </button>
              ))}
              <button onClick={pickImage} className="upload-button">
                <img style={{ width: 50, height: 50, opacity: 0.5 }} src="/assets/upload-icon.png" alt="Upload" />
              </button>
            </div>
            <button onClick={handleCommentSubmit} className="submit-button">
              <span style={{ color: 'white', fontWeight: '900', fontSize: 22 }}>Wyślij</span>
            </button>
          </div>
        </div>
      )}
      <button onClick={handleClick} className="submit-button">
        <span style={{ color: 'white', fontWeight: '900', fontSize: 22 }}>Odpowiedz</span>
      </button>
    </>
  );
}

export default CommentButton;
