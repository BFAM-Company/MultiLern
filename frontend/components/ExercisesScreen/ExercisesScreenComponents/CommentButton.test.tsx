import React from "react";
import { render, fireEvent, waitFor, act } from "@testing-library/react-native";
import CommentButton from "./CommentButton";
import { AxiosContext } from '../../context/AxiosProvider/AxiosProvider';
import { UserDataContext } from '../../context/UserContext/UserContext';
import * as ImagePicker from "expo-image-picker";
import axios from "axios";

jest.mock("@react-native-async-storage/async-storage", () => require("../../../__mocks__/mock-async-storage"));
jest.useFakeTimers()

describe('CommentButton', () => {
    const mockHandleRefresh = jest.fn();
    const mockUserContext = {
      setUserData: jest.fn(),
      userData: {
        id: 1,
        nickname: 'JohnDoe',
        avatar: 'https://example.com/avatar.jpg',
        email: 'john.doe@example.com',
      }
    };
    const mockAxiosContext = {
        publicAxios: {
            post: jest.fn()
        },
        authAxios: {}
    };

    const mockPublicAxios = axios.create();
    const mockAuthAxios = axios.create();

    const setup = () => {
        return render(
            <UserDataContext.Provider value={mockUserContext}>
                <AxiosContext.Provider value={{authAxios: mockPublicAxios, publicAxios: mockPublicAxios}}>
                    <CommentButton id={0} handleRefresh={mockHandleRefresh} />
                </AxiosContext.Provider>
            </UserDataContext.Provider>
        );
    };

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('renders correctly', () => {
        const { toJSON } = setup();
        expect(toJSON()).toMatchSnapshot();
    });

    it('opens the modal on button press', () => {
        const { getByText, getByPlaceholderText } = setup();
        const button = getByText('Odpowiedz');
        fireEvent.press(button);
        expect(getByPlaceholderText('tytuł(opcjonalny)...')).toBeTruthy();
    });

    it('handles text input correctly', () => {
        const { getByText, getByPlaceholderText } = setup();
        fireEvent.press(getByText('Odpowiedz'));
        const titleInput = getByPlaceholderText('tytuł(opcjonalny)...');
        const contentInput = getByPlaceholderText('Rozwiązanie...');

        fireEvent.changeText(titleInput, 'Test Title');
        fireEvent.changeText(contentInput, 'Test Content');

        expect(titleInput.props.value).toBe('Test Title');
        expect(contentInput.props.value).toBe('Test Content');
    });

    it('adds and removes images correctly', async () => {
        const { getByText, getByRole, getByTestId } = setup();
        fireEvent.press(getByText('Odpowiedz'));

        jest.spyOn(ImagePicker, 'requestMediaLibraryPermissionsAsync').mockResolvedValueOnce({ 
            status: ImagePicker.PermissionStatus.GRANTED,
            expires: 'never',
            granted: true,
            canAskAgain: true,});
        jest.spyOn(ImagePicker, 'launchImageLibraryAsync').mockResolvedValueOnce({
            canceled: false,
            assets: [{
                base64: "testbase64",
                uri: "",
                width: 0,
                height: 0
            }]
        });

        const uploadButton = getByTestId('uploadButton');
        await act(async () => {
            fireEvent.press(uploadButton);
        });

        await waitFor(() => {
            expect(getByTestId('image-0')).toBeTruthy();
        });

        const addedImage = getByTestId('image-0');
        fireEvent.press(addedImage);

        await waitFor(() => {
            expect(() => getByTestId('image-0')).toThrow();
        });
    });

    it('submits the comment correctly', async () => {
        const { getByText, getByPlaceholderText } = setup();
        fireEvent.press(getByText('Odpowiedz'));

        const titleInput = getByPlaceholderText('tytuł(opcjonalny)...');
        const contentInput = getByPlaceholderText('Rozwiązanie...');
        fireEvent.changeText(titleInput, 'Test Title');
        fireEvent.changeText(contentInput, 'Test Content');

        const submitButton = getByText('Wyślij');
        
        mockPublicAxios.post = jest.fn().mockResolvedValueOnce({ status: 200 });

        await act(async () => {
            fireEvent.press(submitButton);
        });

        await waitFor(() => {
            expect(mockPublicAxios.post).toHaveBeenCalledWith(
                '/posts/comment',
                {
                    title: 'Test Title',
                    content: 'Test Content',
                    date: expect.any(String),
                    parentPostId: 0,
                    images: [],
                    userId: 1,
                },
                {
                    headers: { 'Content-Type': 'application/json' }
                }
            );
            expect(mockHandleRefresh).toHaveBeenCalled();
        });
    });
});
