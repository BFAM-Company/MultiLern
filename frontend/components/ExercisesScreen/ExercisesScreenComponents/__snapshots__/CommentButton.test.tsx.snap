// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentButton renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "borderRadius": 20,
      "height": 55,
      "margin": 20,
      "opacity": 1,
      "width": 300,
    }
  }
>
  <ViewManagerAdapter_ExpoLinearGradient
    colors={
      [
        4281150775,
        4281150775,
        4284769390,
      ]
    }
    style={
      {
        "alignItems": "center",
        "borderRadius": 20,
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 22,
          "fontWeight": "900",
        }
      }
    >
      Odpowiedz
    </Text>
  </ViewManagerAdapter_ExpoLinearGradient>
</View>
`;
